parquet-hadoop/src/main/java/org/apache/parquet/crypto/AesGcmDecryptor.java

```diff
@@ -0,0 +1,103 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.parquet.crypto;
+
+
+import javax.crypto.Cipher;
+import javax.crypto.SecretKey;
+import javax.crypto.spec.GCMParameterSpec;
+import javax.crypto.spec.SecretKeySpec;
+
+import org.apache.parquet.format.BlockCipher;
+
+import java.io.IOException;
+import java.security.GeneralSecurityException;
+
+class AesGcmDecryptor implements BlockCipher.Decryptor{
+
+  private final SecretKey key;
+  private final byte[] AAD;
+  private final byte[] ivPrefix;
+
+  private static final int GCM_NONCE_LENGTH = AesGcmEncryptor.GCM_NONCE_LENGTH;
+  private static final int GCM_TAG_LENGTH = AesGcmEncryptor.GCM_TAG_LENGTH;
+  private static final int chunkLen = AesGcmEncryptor.chunkLen;
+
+
+  AesGcmDecryptor(byte[] keyBytes, byte[] aad, byte[] ivPrefix) throws IOException {
+    if (null == keyBytes) throw new IOException("Null key bytes");
+    key = new SecretKeySpec(keyBytes, "AES");
+    AAD = aad;
+    this.ivPrefix = ivPrefix;
+    if (null != ivPrefix) {
+      if (ivPrefix.length > GCM_NONCE_LENGTH) throw new IOException("IV prefix length: " + ivPrefix.length);
+    }
+  }
+
+  @Override
+  public byte[] decrypt(byte[] ciphertext)  throws IOException {
+    return decrypt(ciphertext, 0, ciphertext.length);
+  }
+
+  @Override
+  public byte[] decrypt(byte[] ciphertext, int offset, int cLen)  throws IOException {
+    byte[] nonce = new byte[GCM_NONCE_LENGTH];
+    // Get the nonce
+    int noff = 0;
+    int nlen = GCM_NONCE_LENGTH;
+    if (null != ivPrefix) {
```

Comment:
Ok, this is a big subject. Therefore, a long text below.

1. 96 bit is a recommended length for AES-GCM IV, see discussion at
https://crypto.stackexchange.com/questions/41601/aes-gcm-recommended-iv-size-why-12-bytes/41610
eg "*For GCM a 12 byte IV is strongly suggested as other IV lengths will require additional calculations.*"
"*there are additional calculations for IV's other than 96 bits. This is why the original proposal has this recommendation: 96-bit IV values can be processed more efficiently, so that [ed: this] length is recommended for situations in which efficiency is critical.*"

2. For 96-bit IV, the NIST spec (section 8.2) allows two modes: 8.2.1 Deterministic Construction or 8.2.2 RBG-based Construction. 

3. Deterministic IV prefix ("fixed field") "*shall identify the device, or, more generally, the context
for the instance of the authenticated encryption function*". This is not trivial in a single process, and even more challenging in distributed frameworks like Spark ("*no two distinct devices shall share the same fixed field*"). There, in the **writer**, I'm using a RBG-based (random) IV - using all 96 bits as per the NIST section 8.2.2 instructions.

4. In the **reader**, the code can decrypt files with either deterministic or random IV. Thats why we have the iv_prefix. Yes, currently the parquet-mr (and -cpp) **writer** uses only the random option, not setting the `iv_prefix` field. But other Parquet implementations can create files with deterministic IVs, and these files will be readable by current parquet-mr and -cpp.

6. While the random IVs are perfectly fine and follow the NIST spec to the letter - they are even more secure in Parquet, since the vast majority of them is undetectable. Offsets of the pages/ page headers are stored in ColumnMetaData structure - which is encrypted with a column key. Offsets of other modules are stored in the footer, and encrypted with the footer key. Therefore, the offset of most (sometimes all - 1, depending on config) IVs is unknown. And, since they are random, they cant be searched in the file.

7. So, the current design closely  follows the recommendations. However, we can make it a bit more flexible - by allowing IVs with a length other than 96.  This will require an additional optional field `gcm_iv_length` in Thrift `AES_*` structures. While we are at it, we can also allow for GCM and CTR to have a different iv_prefix (setting this as optional field in `AES_*` structures, instead of in `FileCryptoMetaData` struct. Or adding an optional `ctr_iv_prefix` in the AES_GCM_CTR struct). Let me know if you are ok with it, I'll prepare a pull request for this change.

Reply:
lets leave this as is. the changes are minor, and most likely will never be used. 
from the link above - 
"If you supply a larger-than-12-bytes IV then it needs to be "hashed" allowing collisions to happen and raising the risk for (devastating) IV reuse unnecessarily high"
and CTR doesn't really need an IV prefix, can be fully random. plus, like with GCM, most/all IVs are hidden anyway.

Reply:
Ok, I in this case I think we should document that the write path for Java and C++ implementation doesn't support IV prefixes, but read path does. Also, it would be nice to cover this case with tests too to make sure that IV prefix is handled correctly in the read path.

I'm not sure how likely it is that one would like to specify different IV length, we can open a Jira for that if you think.

Reply:
Sure, I'll document this, and add a tester.
Custom IV length seems to be unnecessary. But currently we have a bit of ambiguity of what the iv_prefix applies to - GCM or CTR or both. I'll open a Jira that renames this parameter, and, per your earlier suggestion, moves it to the Algorithms structures (from the  FileCryptoMetaData structure).